doc/qos/scheduler/SONiC_QoS_Scheduler_Shaper.md

```diff
@@ -0,0 +1,471 @@
+
+Max shaping configuration can be applied to any front panel port.
+
+### 1.1.4 Warm Boot Requirements
+Scheduling and shaping functionalities continue to work across warm reboot. 
+
+- To support planned system warm boot.
+- To support SWSS docker warm boot.
+
+
+## 1.2 Design Overview
+
+
+### 1.2.1 Basic Approach
+The QoS Orchestration agent is used for handling QoS scheduling and shaping configuration.  The scheduler object is defined as a profile to control several scheduling attributes such as:
+- scheduling type (SP, WRR, DWRR)
+- scheduling weight
+- shaper min/max bandwidth rate
+
+Scheduler APIs include create/remove scheduler profiles and set/get scheduler attributes.  A scheduler profile can be mapped to a queue or a port.
+
+### 1.2.2 Container
+No new containers are introduced. Functionality is added to the existing SWSS container. 
+
+### 1.2.3 SAI Overview
+The existing SAI scheduler, queue and port APIs shall be used to support this feature and they are available on SAI github:
+
+https://github.com/opencomputeproject/SAI/blob/master/inc/saischeduler.h<br>
+https://github.com/opencomputeproject/SAI/blob/master/inc/saiqueue.h<br>
+https://github.com/opencomputeproject/SAI/blob/master/inc/saiport.h<br>
+
+
+The details of the changes will be discussed in the design section below.
+
+
+# 2 Functionality
+Refer to section 1
+## 2.1 Target Deployment Use Cases
+Any network that requires differentiation of traffic types and services by managing bandwidth resources.  The typical use case is for protection of real-time and high priority data.
+
+Data centers use traffic shaping to maintain service level agreements for the variety of applications and the many tenants hosted on a shared physical network.
+
+
+## 2.2 Functional Description
+Refer to section 1.1
+
+## 2.3 Limitations
+QoS scheduling and shaping 
+- Configuration is only supported on physical interfaces and physical interface egress queues.
+- Configuration is not supported on VLAN and port-channel interfaces.
+
+# 3 Design
+## 3.1 Overview
+The design overview at a high level:
+
+### 3.1.1 Creating a Scheduler Profile
+1) QoS scheduler and shaper configuration profiles are parsed from the JSON file and stored in CFG_SCHEDULER_TABLE in the configuration database.
+2) The QoS Orchestration Agent subscribes to notifications from the CFG_SCHEDULER_TABLE and parses the input parameters (scheduling type, weight, min/max bandwidth rate) of the scheduler profile.
+3) create_scheduler SAI API is invoked to create a scheduler profile with the given input parameters. 
+4) The scheduler ID returned from the SAI create_scheduler API is associated with the scheduler profile and the profile data.
+
+### 3.1.2 Applying the Scheduler Profile to Queues
+
+1) The QoS Scheduler can be applied to a queue (or set of queues) by setting the QoS scheduler profile name to the scheduler attribute of the queue (or queue range) of an interface in the CFG_QUEUE_TABLE.  
```

Comment:
Can we support scheduler on CPU queues? SAI provides a way to get the CPU queue OIDs switch->cpu_port->port_queue_list(). How do we expose this to SONiC user?

Reply:
Yes we can support it on CPU queues.  I think the most consistent way is to go through the COPP file so that trap rate limiters and CPU queue shaping can be coordinated in one place.   It is definitely useful but out of the scope for this HLD.

Reply:
Sure. Please update the scope of the HLD to say that it only refers to front panel ports and their queues.